Add typical carbon year (TCY) functionality

ssempervirens opened the pull request:
This PR adds functionality to calculate a Typical Carbon Year (TCY) profile for any given year using 3 years of historical MOER data. 

Changes:
- Add TCY implementation
- Add testing 
- Update documentation

The implementation takes historical MOER data from past 3 years, aggregates by month, hour, and day type, converts timezones, and returns hourly values. 

The tests I wrote pass, and I also tested the implementation manually with my login credentials.

ssempervirens committed:
Add Typical Carbon Year (TCY) calculation functionality

ssempervirens committed:
refine readme text

nsteins commented on watttime/tcy.py:
```diff
@@ -0,0 +1,109 @@
+class TCYCalculator:
+    """Calculate Typical Carbon Year profiles from historical MOER data using a 3-year lookback period"""
+    
+    def __init__(self, username: str, password: str, config: TCYConfig):
```
```suggestion
    def __init__(self, username: Optional[str] = None, password: Optional[str] = None, config: TCYConfig):
```

This lets you skip specifying username and password, which will use the env vars when `None`s are passed to `WattTimeHistorical`

nsteins reviewed (commented):
This looks great to me, I'd just request a few small changes:

- I think the `TCYConfig` may be an unnecessary object to expose to end users. Since the config only applies to the `TCYCalculator` object, you could just pass the arguments directly into init, unless there is a specific reason you wanted to use a Config object. That could look like `TCYCalculator(region, timezone, holidays, username, password)`
- I would definitely make username and password optional init args, that use the env variables if no value is passed, just like the API classes.

Thanks for adding tests and documentation!!

sam-watttime commented on watttime/tcy.py:
```diff
@@ -0,0 +1,109 @@
+    """Calculate Typical Carbon Year profiles from historical MOER data using a 3-year lookback period"""
+    
+    def __init__(self, username: str, password: str, config: TCYConfig):
+        self.wt_client = WattTimeHistorical(username, password)
```
can this class just be a child of WattTimeHistorical?

nsteins replied:
I had suggested making this independent, since `WattTimeHistorical` is a child of `WattTimeAPI` and this isn't really an API object, but rather a utility calculation that uses API data.

ssempervirens committed:
Refactor TCY implementation to simplify interface

- Remove TCYConfig class and move parameters directly to TCYCalculator
- Make username/password optional with env var fallback
- Fix deprecation warnings in pandas date handling
- Update tests and documentation

nsteins commented on README.md:
```diff
@@ -144,3 +144,38 @@ wt.region_from_loc(
+from watttime import TCYConfig, TCYCalculator
+
+# Configure TCY calculation
+config = TCYConfig(
```
You can remove the TCYConfig section from the readme

nsteins commented on watttime/tcy.py:
```diff
@@ -0,0 +1,125 @@
+        self.password = password or os.getenv("WATTTIME_PASSWORD")
+        
+        # Initialize API client with credentials
+        self.wt_client = WattTimeHistorical(self.username, self.password)
```
```suggestion
        self.wt_client = WattTimeHistorical(username, password)
```

You can delete the previous lines and just pass the arguments  straight to `WattTimeHistorical`. If they default to `None`, then we pass `None` to `WattTimeHistorical`, which then looks at the getenv. It's not strictly necessary, but it keeps us from repeating the same functionality twice in the code base, so future updates are easier to manage if it's only handled in one location.

nsteins reviewed (approved):
Looks good! made two minor suggestions, but this looks great!

Only note, is let's keep an eye on time to run tests - since the test now grabs 3 years of data to calculate the TCY. But looking at it right now, it doesn't look like a problem

ssempervirens committed:
nat's requested changes

ssempervirens committed:
add holidays to import

geoffhancock commented on README.md:
```diff
@@ -144,3 +144,33 @@ wt.region_from_loc(
+The returned tcy_profile is a pandas series with hourly timestamps in the specified timezone and corresponding MOER values (in lbs CO2/MWh). The profile provides hourly values that represent typical grid emissions patterns for the region. This can be useful for:
+* Building energy modeling and optimization
+* Understanding typical emissions patterns in a region
+* Planning load shifting strategies
```
This data is not good for planning load shifting strategies... since it does not exhibit realistic MOER variation that occurs in a real day, due to the smoothing that happens from averaging. If one were to use this data to estimate the potential carbon reduction opportunity of doing AER, the result would be significantly lower than what a real-time optimization would produce with actual MOERs. I suggest we remove this list item and add a caveat in this section that says something like: "this data has a lower variance than real-time MOER data, so it should not be used to estimate the carbon reduction opportunity from load-shifting/AER."

geoffhancock commented on README.md:
````diff
@@ -144,3 +144,33 @@ wt.region_from_loc(
+```
+
+The returned tcy_profile is a pandas series with hourly timestamps in the specified timezone and corresponding MOER values (in lbs CO2/MWh). The profile provides hourly values that represent typical grid emissions patterns for the region. This can be useful for:
+* Building energy modeling and optimization
````
suggestion:
Building energy modeling and optimization -> Estimating the relative carbon impact of building design choices 

For the reasons below, we probably don't want to use "optimization" here.
Also, this data is more applicable to "static" design choices, where it is the rough/high-level patterns that matter more than the hour-to-hour variations of a particular day.

geoffhancock commented on README.md:
````diff
@@ -144,3 +144,33 @@ wt.region_from_loc(
 ```
+
+### Calculating a Typical Carbon Year (TCY)
+The SDK provides functionality to calculate a Typical Carbon Year (TCY) profile, which represents typical MOER patterns over a year by combining three years of historical data. A TCY accounts for seasonal variations, daily patterns, and differences between weekdays and weekends/holidays:
````
by combining **the most recent three** years of historical data

geoffhancock replied:
I think this clarification may avoid some potential confusion

ssempervirens committed:
update readme in response to geoff's comments. clarify that we are using 3 years of historical data, removed load shfiting from use cases, and added caveat about variance

geoffhancock reviewed (approved):
Thanks for making the suggested edits to the readme!
